### DISCUSSION TITLE
Any reason why `/api/user/info` returns such limited information?

### DISCUSSION THREAD

**Discussion by mikew** [authelia/authelia#6913] [General]
Asking because I have some integrations that require getting the person's full name, groups, and email of an Authelia user from just an Authelia cookie.

Up to now, I've been using the `/api/authz/auth-request`, as it's the most straightforward to call, and returns the info needed. It's only now that I'm upgrading to 4.38 final that I realize it's harder to share these integrations because:

- That route might not exist
- That route might exist but not use `CookieSession`

And accounting for all these makes things more difficult than it feels like they have to be by:

- Telling the person using the integration to "make sure your Authelia installation has an `AuthRequest` Authz Endpoint with a `CookieSession` auth strategy"
- Programming in support for passing the potentially different authz routes
- Programming in support for different authz endpoint implementations

If the API could give that info from `/user/info` it could save people writing and using Authelia integrations effort.

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Based on the context provided and the details of your question, it seems you're looking for a way to simplify the process of retrieving user information (full name, groups, and email) from Authelia, especially considering the potential variability in the availability and configuration of authorization endpoints (`/api/authz/auth-request`) and authentication strategies (like `CookieSession`).

As of the information available up to Authelia version 4.38, the `/api/user/info` endpoint does not directly provide the detailed user information (full name, groups, and email) you're seeking. This endpoint primarily returns information about the authentication method the user has configured (e.g., `totp`, `webauthn`) rather than detailed identity information.

However, Authelia supports OpenID Connect (OIDC), which can be used to retrieve detailed user information through the `/userinfo` endpoint once a user is authenticated. OIDC is a more standardized way to achieve what you're looking for, as it's designed to share user information securely between authentication systems and applications.

To use OIDC for retrieving user information, you would:

1. Ensure your Authelia instance is configured to use OIDC.
2. Authenticate the user via Authelia's OIDC flow.
3. Use the access token obtained during authentication to make a request to the `/userinfo` endpoint.

This approach would require your integrations to support OIDC flows, but it would standardize the way you retrieve user information across different setups and configurations of Authelia.

Regarding the direct API call to `/user/info` for detailed user information as you described, this is not currently supported as per the provided context. It might be worth considering contributing to Authelia or requesting this feature directly in the Authelia GitHub repository discussions or issues, as it could benefit other users with similar integration needs.

For now, leveraging OIDC and its `/userinfo` endpoint is the recommended approach to obtain detailed user information in a standardized manner.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
So generally these endpoints are used for the official frontend and we encourage most integrations to occur via OpenID Connect 1.0 or the Remote-* response headers which can be automatically injected into the request by most proxies transparently. The endpoints being used are not specifically intended for integration. We operate with a least information needed approach to as many things as possible, I'd rather not permit the use of resource owner credentials (cookies or username/password in this instance) being used to access things they have no need for.

We are likely to add an endpoint for an implementer that leverages the Client Credentials Grant to allow this kind of integration to occur more programmatically in a secure way which has specific intention. If you're interested in being involved I'm more than willing to let you know once this feature is ready to test.

**Reply by mikew** [reply]
> the Remote-* response headers which can be automatically injected into the request by most proxies transparently

I am using those headers as a quick and dirty method, but the thing that worries me is that someone with access to the docker network (or in some cases any machine on the network) can easily form a request with the `Remote-*` headers directly to the service, and the integration would just assume they are who they say they are, with no actual verification happening. My goal with accessing user information via the cookie is to provide something a little more.

I would love to use OpenID Connect, but in one case it's not feasible, and in another case, the integration exists to smooth over the apps own implementation of OIDC.

> We are likely to add an endpoint for an implementer that leverages the Client Credentials Grant

At first glance it looks like that might help, but what about getting information about the current user? I've only just read about Client Credentials Grant, and it looks like the flow would be:

- Register client id + secret with Authelia
- Integration uses that id + secret to get an Authelia token

But that Authelia token isn't associated with a user, so I don't know what `/api/oidc/userinfo` would return.

**Reply by james-d-elliott** [reply]
Specifically how the Client Credentials Flow helps is you can grant specific scopes and audiences which allow the token to do various things. 

Some example ideas I've been thinking over are accessing a `/api/oidc/userinfo/<user_hint>/<username>` endpoint. Where `<user_hint>` is something only Authelia and the backend application knows (via a header) about the user such as a UUID which is associated per-user and per-origin i.e. the origin and username are passed through a HMAC to generate the unique value or it's just a UUID per those two fields.

Access Servers can have lots of endpoints protected by OAuth 2.0, the Userinfo endpoint is just a specifically defined one by the specifications of OpenID Connect 1.0.
